RFC add babeld-auto-gw-mode

spiccinini opened the pull request:
By default babeld will redistribute all the routes installed even if they "don't work". For example when the internet provider use DHCP and the service is not working but interface is up, the route is installed but not working and babeld will anounce the non working route to the network.

This package provides a solucion using watchping hooks that adds routes with a special protocol number (7) when the WAN port has a working internet access and removes this route when the internet connection is not working as detected by watchping.

Fixes  #93 #822


Open questions: 
* Is it ok to use protocol 7?
* ~~Should the route be installed with its `src` attribute?~~ No
* ~~Should we remove the default gw route when the internet connection is not working and replace it with a route to only the src address? This would fix #822 any other ideas?~~ DONE

~~TODO:  add ipv6 support~~ Not for this PR

G10h4ck commented on packages/babeld-auto-gw-mode/Makefile:
```diff
@@ -0,0 +1,22 @@
+define Package/$(PKG_NAME)
+  SECTION:=utils
+  CATEGORY:=Utilities
+  TITLE:=Bebeld auto Internet gateway module.
```
Babeld ;)

spiccinini replied:
Thanks

G10h4ck commented on packages/babeld-auto-gw-mode/Makefile:
```diff
@@ -0,0 +1,22 @@
+  CATEGORY:=Utilities
+  TITLE:=Bebeld auto Internet gateway module.
+  MAINTAINER:=Santiago Piccinini <spiccinini@altermundi.net>
+  DEPENDS:=+libubus-lua +watchping +ip +lime-proto-babeld
```
`+ip` toma el de busybox o el de iproute2 ?

spiccinini replied:
you are right, ip uses iproute2, I am removing it.

G10h4ck commented on packages/babeld-auto-gw-mode/files/etc/watchping/wan-ok.d/babeld-gw:
```diff
@@ -0,0 +1,21 @@
+
+for _, route4 in ipairs(wan4_status.route) do
+    if route4.target == "0.0.0.0" and route4.mask == 0 then
+        --TODO: maybe add also src config?
```
As this route will not really used to route actual packets, but just for babeld to read it and spread to the network there is no advantage in specifying the source address

spiccinini replied:
Thanks for the clarification, I will remove the comment

ilario commented:
Amazing SAn, thanks!!!!
I cannot test right now, maybe @amuuza?

amuuza commented:
Thank you @ilario for letting me know about this, I did not notice.
Oh, but I am reading it does not fix #822... I'm afraid I do not understand what this fix is about!

In my previous tests, when an Internet gateway of a router failed, any other router stopped using it, and they automatically (after several minutes) switched to a working Internet gateway. Any other router but itself. The router with the failing Internet gateway did not automatically switch to a working one, so all its clients were not able to reach the Internet unless the router was manually rebooted.
Has anything been fixed of this situation?
I would be very happy to do a test if it fixed #822

spiccinini commented:
Hi! To fix #822 what do you think about Increasing the metric, by a very large amount,  of the default gw when it is not working so that this route is not used if there is other default route distributed by other node. 
Do you think this is a viable solution?

spiccinini commented:
> Hi! To fix #822 what do you think about Increasing the metric, by a very large amount, of the default gw when it is not working so that this route is not used if there is other default route distributed by other node.
> Do you think this is a viable solution?

@G10h4ck what do you think? I will try this idea and report back but I would love a confirmation that it is sound.

spiccinini commented:
> Hi! To fix #822 what do you think about Increasing the metric, by a very large amount, of the default gw when it is not working so that this route is not used if there is other default route distributed by other node.
> Do you think this is a viable solution?

I have implemented and tested this using the new qemu cloud work from guido (https://github.com/libremesh/lime-packages/pull/813). It is working as expected :)  
Note that the default watchping configuration takes up to 120 seconds to refresh the interfaces so, it is not instantaneous.

@amuuza can you test it?!

spiccinini commented:
Note that ipv6 support is still not done.

amuuza commented:
Hi @spiccinini 
Yes, I can, but how?
I guess I should follow the compiling instructions at https://libremesh.org/development.html but use
`src-git libremesh https://github.com/spiccinini/lime-packages.git;master` instead of
`src-git libremesh https://github.com/libremesh/lime-packages.git;master`
Is that correct? Any other change?

ilario commented on packages/babeld-auto-gw-mode/Makefile:
```diff
@@ -0,0 +1,22 @@
+define Package/$(PKG_NAME)
+  SECTION:=utils
+  CATEGORY:=Utilities
+  TITLE:=Bebel auto Internet gateway module.
```
Bebel typo also here :)

spiccinini replied:
Thanks!

ilario commented:
> I guess I should follow the compiling instructions at https://libremesh.org/development.html but use
> `src-git libremesh https://github.com/spiccinini/lime-packages.git;master`

Quite there! In that same line you also have to replace `master` by `babeld-auto-gw-mode` which is the name of the branch mentioned at the top of this page.
Following the guide on the development page you should also refresh the feeds.
Finally, in Menuconfig, you'll have to select the new package `Utilities/babeld-auto-gw-mode` and compile.
Let us know pls!

amuuza commented:
Thank you @ilario. I'm getting some problems.

I did
`scripts/feeds update -a`
It seemed to work ok but then I was asked for a Github username and password, I don't know why.

Then I did
`scripts/feeds install -a`
and got many warnings. See [here](https://0bin.net/paste/lT9oRBhj#7qV6cHr93Bji-PS9EUtCk1M5wj92BqgJkikofNV2U3f).

And menuconfig misses many libremesh items.

ilario commented:
Weird. Try deleting the feeds/ directory and run again the update.
Also, paste here the content of your feeds.conf

amuuza commented:
Thanks. Everything seems ok now, I don't know why. It is now compiling.
The only difference I have found is that now I removed the break before
`src-git libremesh https://github.com/spiccinini/lime-packages.git;babeld-auto-gw-mode`
I don't know if that could be the reason, otherwise I did some mistake I cannot identify.

I'll let you know when the tests are done.

amuuza commented:
```
 ISP-A                     ISP-C
   |                         |
RouterA <--> RouterB <--> RouterC
```
RouterA, RouterB and RouterC are wirelessly meshed.
RouterA WAN port is connected to ISP-A to reach the Internet.
RouterC WAN port is connected to ISP-C to reach the Internet.

If RouterB is reaching the Internet through RouterA, and ISP-A loses its Internet connectivity, then both RouterA and RouterB obviously lose their Internet connectivity. After 8 minutes approximately, Router B switches its Internet gateway automatically to ISP-C.

However, RouterA will not switchover automatically.
I waited for more than 10 minutes.
Then I rebooted it and it did not switchover either.
It only did the switch after both pulling off its WAN cable and rebooting it.

spiccinini commented:
@amuuza thanks for testing!! So it is not working as expected for you. Can you check that you have correctly installed the new package babeld-auto-gw-mode in RouterA ? (with opkg -l | grep babel)
Also please check if watchping is running with logread | grep watchping, you should see some mesages. 
If you can join the chatroom at https://www.libremesh.org/communication.html we can debug this faster :)

amuuza commented:
Ooooh, I forgot to include it with Menuconfig!! Ilario told me so, but I forgot!!

I'll do everything again and let you know.

By the way, isn't there any way to do the tests through software? Maybe GNS3?

amuuza commented:
I reinstalled RouterA, now with the right image (I hope so).

I waited for more than 10 minutes and there was no automatic switchover. It only worked when I manually rebooted the router.

There is a just a little improvement: this time it only required a manual reboot to work, there was no need to take out the active (but without Internet) WAN cable to make it work.

While waiting for the switchover I entered your suggestions:
```

root@MV-aa8ebe:~# opkg list | grep babel
babeld - 1.9.2-1
babeld-auto-gw-mode - git-21.043.62159-8f372ae-1
lime-proto-babeld - git-21.043.62159-8f372ae
shared-state-babeld_hosts - 2021-02-07-1612741008
root@MV-aa8ebe:~# 
root@MV-aa8ebe:~# logread | grep babel
Sun Dec  6 07:31:33 2020 daemon.notice netifd: Interface 'lm_net_eth0_1_babeld_if' is enabled
Sun Dec  6 07:31:33 2020 daemon.notice netifd: Interface 'lm_net_eth0_1_babeld_if' is setting up now
Sun Dec  6 07:31:33 2020 daemon.notice netifd: Interface 'lm_net_eth0_1_babeld_if' is now up
Sun Dec  6 07:31:33 2020 daemon.notice netifd: Interface 'lm_net_eth0_2_babeld_if' is enabled
Sun Dec  6 07:31:33 2020 daemon.notice netifd: Interface 'lm_net_eth0_2_babeld_if' is setting up now
Sun Dec  6 07:31:33 2020 daemon.notice netifd: Interface 'lm_net_eth0_2_babeld_if' is now up
Sun Dec  6 07:31:34 2020 daemon.notice netifd: Interface 'lm_net_eth0_babeld_if' is enabled
Sun Dec  6 07:31:34 2020 daemon.notice netifd: Interface 'lm_net_eth0_babeld_if' is setting up now
Sun Dec  6 07:31:34 2020 daemon.notice netifd: Interface 'lm_net_eth0_babeld_if' is now up
Sun Dec  6 07:31:34 2020 daemon.notice netifd: Interface 'lm_net_eth0_1_babeld_if' has link connectivity
Sun Dec  6 07:31:34 2020 daemon.notice netifd: Interface 'lm_net_eth0_2_babeld_if' has link connectivity
Sun Dec  6 07:31:34 2020 daemon.notice netifd: Interface 'lm_net_eth0_babeld_if' has link connectivity
Sun Dec  6 07:31:38 2020 daemon.err babeld[1921]: Warning: couldn't determine channel of interface eth0-1_17.
Sun Dec  6 07:31:39 2020 daemon.err babeld[1921]: Warning: couldn't determine channel of interface eth0_17.
Sun Dec  6 07:31:39 2020 daemon.err babeld[1921]: Warning: couldn't determine channel of interface eth0-2_17.
Sun Dec  6 07:31:48 2020 daemon.notice netifd: Interface 'lm_net_wlan1_mesh_babeld_if' is enabled
Sun Dec  6 07:31:48 2020 daemon.notice netifd: Interface 'lm_net_wlan1_mesh_babeld_if' is setting up now
Sun Dec  6 07:31:48 2020 daemon.notice netifd: Interface 'lm_net_wlan1_mesh_babeld_if' is now up
Sun Dec  6 07:31:49 2020 daemon.notice netifd: Interface 'lm_net_wlan0_mesh_babeld_if' is enabled
Sun Dec  6 07:31:49 2020 daemon.notice netifd: Interface 'lm_net_wlan0_mesh_babeld_if' is setting up now
Sun Dec  6 07:31:49 2020 daemon.notice netifd: Interface 'lm_net_wlan0_mesh_babeld_if' is now up
Sun Dec  6 07:31:49 2020 daemon.err babeld[1921]: Warning: couldn't determine channel of interface wlan1-mesh_17.
Sun Dec  6 07:31:49 2020 daemon.err babeld[1921]: send: Address not available
Sun Dec  6 07:31:49 2020 daemon.err babeld[1921]: Warning: couldn't determine channel of interface wlan0-mesh_17.
Sun Dec  6 07:31:49 2020 daemon.notice netifd: Interface 'lm_net_wlan1_mesh_babeld_if' has link connectivity
Sun Dec  6 07:31:49 2020 daemon.err babeld[1921]: send: Address not available
Sun Dec  6 07:31:49 2020 daemon.notice netifd: Interface 'lm_net_wlan0_mesh_babeld_if' has link connectivity
Sun Dec  6 07:31:58 2020 daemon.notice netifd: Interface 'lm_net_wlan1_mesh_babeld_if' is now down
Sun Dec  6 07:31:58 2020 daemon.notice netifd: Interface 'lm_net_wlan1_mesh_babeld_if' is disabled
Sun Dec  6 07:31:58 2020 daemon.notice netifd: Interface 'lm_net_wlan1_mesh_babeld_if' has link connectivity loss
Sun Feb 14 21:50:00 2021 cron.info crond[1461]: USER root pid 3941 cmd ((sleep $(($RANDOM % 120)); shared-state sync babeld-hosts &> /dev/null)&)
Sun Feb 14 21:55:00 2021 cron.info crond[1461]: USER root pid 5002 cmd ((sleep $(($RANDOM % 120)); shared-state sync babeld-hosts &> /dev/null)&)
Sun Feb 14 22:00:00 2021 cron.info crond[1461]: USER root pid 5817 cmd ((sleep $(($RANDOM % 120)); shared-state sync babeld-hosts &> /dev/null)&)
root@MV-aa8ebe:~#

```

@spiccinini, if you want me to test something while being on IRC, let me know which UTC times are ok with you.

spiccinini commented:
> Ooooh, I forgot to include it with Menuconfig!! Ilario told me so, but I forgot!!
> 
> I'll do everything again and let you know.
> 
> By the way, isn't there any way to do the tests through software? Maybe GNS3?

For testing with software I am using full virtualization with qemu as explained in https://github.com/libremesh/lime-packages/blob/master/TESTING.md#development-with-qemu-virtual-machine

In particular to test this I used this PR https://github.com/libremesh/lime-packages/pull/813 that allows to create multiple mesh clouds interconected. But I had to do some manual things to do the WAN/internet access.

spiccinini commented:
> I reinstalled RouterA, now with the right image (I hope so).
> 
> I waited for more than 10 minutes and there was no automatic switchover. It only worked when I manually rebooted the router.
> 
> There is a just a little improvement: this time it only required a manual reboot to work, there was no need to take out the active (but without Internet) WAN cable to make it work.

OK so it is not working as expected yet.


> While waiting for the switchover I entered your suggestions:
> 
> ```
> 
> root@MV-aa8ebe:~# opkg list | grep babel
> babeld - 1.9.2-1
> babeld-auto-gw-mode - git-21.043.62159-8f372ae-1
> lime-proto-babeld - git-21.043.62159-8f372ae
> shared-state-babeld_hosts - 2021-02-07-1612741008

Ok this looks good :)

>
> root@MV-aa8ebe:~# 
> root@MV-aa8ebe:~# logread | grep babel
> Sun Dec  6 07:31:33 2020 daemon.notice netifd: Interface 'lm_net_eth0_1_babeld_if' is enabled
> root@MV-aa8ebe:~#
> ```

Can you grep using `watchping` instead of `babel`?  

> @spiccinini, if you want me to test something while being on IRC, let me know which UTC times are ok with you.

I am usually online from 10 to 20 UTC-3. 

Thanks for all the work!

amuuza commented:
Oops, sorry, I did `grep babel` instead of `grep watchping`.

> I am usually online from 10 to 20 UTC-3.

Cool. Maybe later today or tomorrow, I'll let you know.

spiccinini commented:
I've found a bug! I have already fixed it. 

@amuuza You will have to run the feed update, please check if there is  an error delete the feeds/libremesh folder and re run the feeds update.

amuuza commented:
Yes, now it does work!! Thank you!

In one or two minutes RouterA was back online, despite having its no-internet-WAN connected.

amuuza commented:
I think this is an important step for Internet sharing.

Now I have some questions about metrics and monitoring... but I'll ask them in the list.

spiccinini commented:
Thanks to @amuuza and @ilario I think this package is ready to be merged now

spiccinini commented:
Sorry, ipv6 support is not done yet, I will work on that now.

spiccinini commented:
Travis build is failing with 
```
anyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
```
I don't know if there is something we can do other than changing the docker registry.

nicopace commented:
> Travis build is failing with
> 
> ```
> anyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
> ```
> 
> I don't know if there is something we can do other than changing the docker registry.

seems there are some options here: https://docs.travis-ci.com/user/docker/

rallep71 commented:
Hello, 
Question, it is an important extension of Libremesh, when will it go further? 
Thanks already to all who made it possible.

spiccinini commented:
I have reverted the ipv6 support as it may be problematic in some scenarios. I propose that we merge this with ipv4 and then if there is need of ipv6 support we can implement it (I reverted the ipv6 patch so we can have it at hand in the git history).

G10h4ck commented:
> > Hi! To fix #822 what do you think about Increasing the metric, by a very large amount, of the default gw when it is not working so that this route is not used if there is other default route distributed by other node.
> > Do you think this is a viable solution?
> 
> @G10h4ck what do you think? I will try this idea and report back but I would love a confirmation that it is sound.

AFAIR babeld doesn't import the metric associated to the route in the kernel routing table, as metric for the kernel have different meaning then metric for babeld, so most probably this would not work as easy as it seems

amuuza commented:
I see I've been asked for a review, but I can't hardly read any code. What I have done is testing it, if that's enough to give an approval review let me know and I'll do it.

spiccinini commented:
> I see I've been asked for a review, but I can't hardly read any code. What I have done is testing it, if that's enough to give an approval review let me know and I'll do it.

Given that we need 2 reviews to merge, in my opinion, testing is  a valid review.

nicopace commented on packages/babeld-auto-gw-mode/Readme.md:
```diff
@@ -0,0 +1,16 @@
+By default babeld will redistribute all the routes installed even if they "don't work". For example
+when the internet provider use DHCP and the service is not working but interface is up, the route is
+installed but not working and babeld will anounce the non working route to the network and also
+this route will be used by this node to route the packets so also all its clients won't have inernet.
```
s/internet/internet

nicopace commented on packages/babeld-auto-gw-mode/files/etc/hotplug.d/lime-config/babeld-auto-gw.lua:
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/lua
```
Is there any concrete use of the `lime-config` hotplug.d subdirectory?
It looks like a grouping that is not documented, and it is not clear for me if it should be there or not.

spiccinini replied:
Indeed lime-config hotplug.d  documentation is missing, it was introduced in early 2019. 
hotplug.d for lime-config is similar to having a generic `run_asset` configured with `ATCONFIG` but it is enabled just by puting the script in the directory.

nicopace reviewed (approved):
did some comments... looks pretty straightforward. haven't tested it out myself, but have read each line.
I am a little worried how the ipv6 conversation is harder... @altergui may be you want to contribute to it (as I remember you are invested in ipv6 support).
Thanks @spiccinini  for working on this!

amuuza reviewed (approved):
I tested it. It works.

AndyMcSchopf commented:
Wow, just saw this in the mailinglist. Great to see progress and a fix for this feature! Will also test it the next days. Thanks for the work!!!
